src/arc.rs

```diff
@@ -168,30 +168,49 @@ impl<T: Copy> From<&[T]> for Arc<[T]> {
 }
 
 #[allow(clippy::fallible_impl_from)]
-impl<T> From<Box<T>> for Arc<T> {
+impl<T: ?Sized> From<Box<T>> for Arc<T> {
     #[inline]
     fn from(b: Box<T>) -> Arc<T> {
-        let align =
-            std::cmp::max(mem::align_of::<T>(), mem::align_of::<AtomicUsize>());
-
-        let rc_width = std::cmp::max(align, mem::size_of::<AtomicUsize>());
-
         unsafe {
-            let dst_layout = Layout::new::<ArcInner<T>>();
-            let dst = alloc(dst_layout);
-            assert!(!dst.is_null(), "failed to allocate Arc");
+            let src = Box::into_raw(b);
+            let value_size = std::mem::size_of_val(&*src);
+            let value_layout = Layout::for_value(&*src);
+
+            let align = std::cmp::max(value_layout.align(), mem::align_of::<AtomicUsize>());
+            let rc_width = std::cmp::max(align, mem::size_of::<AtomicUsize>());
+            let unpadded_size = rc_width.checked_add(value_size).unwrap();
+            let size = (unpadded_size + align - 1) & !(align - 1);
+            let dst_layout = Layout::from_size_align(size, align).unwrap();
+            let dst_thin = alloc(dst_layout);
+            assert!(!dst_thin.is_null(), "failed to allocate Arc");
+
+            // If *mut T is a fat pointer, this will copy the slice length
+            // from the source box, and overwrite the pointer to point to the
+            // newly allocated memory. Otherwise, if *mut T is a thin pointer,
+            // this will overwrite the entire pointer, and dst will be
+            // identical to dst_thin, save for its type.
+            let mut dst = src as *mut ArcInner<T>;
+            #[allow(trivial_casts)]
+            ptr::write(
+                &mut dst as *mut _ as *mut *mut u8,
+                dst_thin as *mut u8,
+            );
```

Comment:
This is the part that I'm stumbling a bit over. I'm a bit worried that this might break if the internal layout for fat pointers changes. The reason I went with the fatten method from https://users.rust-lang.org/t/construct-fat-pointer-to-struct/29198/9 was because @dtolnay provided it as a way to fatten things without assuming the memory layout. I think I want to avoid this if a later version of Rust changes the internal layout for fat pointers and causes this to break. I'm having trouble knowing if this will also break.

Reply:
Ah, good point. I can make use of that to write a narrower `impl<T> From<Box<[T]>>`.